src/libraries/Microsoft.PowerFx.Core/Public/Values/CollectionTableValue.cs

```diff
@@ -19,23 +20,51 @@ namespace Microsoft.PowerFx.Types
     public abstract class CollectionTableValue<T> : TableValue
     {
         // Required - basic enumeration. This is the source object. 
-        private readonly IEnumerable<T> _enumerator; // required. supports enumeration;
+        private IEnumerable<T> _enumerator; // required. supports enumeration;
+
+        // Async implementation
+        private IAsyncEnumerable<T> _asyncEnumerable;        
 
         // Additional capabilities. 
-        private readonly IReadOnlyList<T> _sourceIndex; // maybe null. supports index. 
-        private readonly IReadOnlyCollection<T> _sourceCount; // maybe null. supports count;
-        private readonly ICollection<T> _sourceList; // maybe null. supports mutation;
+        private IReadOnlyList<T> _sourceIndex; // maybe null. supports index. 
+        private IReadOnlyCollection<T> _sourceCount; // maybe null. supports count;
+        private ICollection<T> _sourceList; // maybe null. supports mutation;
 
         public CollectionTableValue(RecordType recordType, IEnumerable<T> source)
           : this(IRContext.NotInSource(recordType.ToTable()), source)
         {
             RecordType = recordType;
         }
 
+        public CollectionTableValue(RecordType recordType, bool inMemory, IAsyncEnumerable<T> source)
```

Comment:
>bool inMemory, [](http://example.com/codeflow?start=59&length=14)

why do we need this? We need to be very clear about the invariants around this new flag.

Reply:
indirectly, to consider capabilities